### PR TITLE
Auto-close blank popups when download is complete

### PR DESCRIPTION
**Pull request by jamespearce2006**
Fix for issue #1031

Automatically close blank popups that appear when downloads are started in a separate window, once the download is done (complete, failed or cancelled).

Test the fix here:
https://jsfiddle.net/yLk7d3z8/

### CODE REVIEW

File: CefSharp.Core/Internals/DownloadAdapter.cpp [hunk 1]
```diff
@@ -38,6 +38,12 @@ namespace CefSharp
         void DownloadAdapter::OnDownloadUpdated(CefRefPtr<CefBrowser> browser, CefRefPtr<CefDownloadItem> download_item,
             CefRefPtr<CefDownloadItemCallback> callback) 
         {
+
```
Comment (jankurianski) [inline CefSharp.Core/Internals/DownloadAdapter.cpp]: Remove extra blank line.

**Comment by amaitland** [comment]
If we merge this into `master`, I'd probably remove it in `2272` and give the user control to close their own browsers since we've introduced a wrapper. (I don't think we expose the close method yet, we should though).

**Comment by jankurianski** [comment]
Ah, you're talking about the working being done in #1030? So the `IDownloadHandler` method could turn into:

    bool OnDownloadUpdated(CefSharpBrowserWrapper browser, DownloadItem downloadItem)

And then users can do the `browser.Close()` if they chose?  I think :+1:

**Comment by amaitland** [comment]
> And then users can do the browser.Close() if they chose?

Yep, that's the one.

File: CefSharp.Core/Internals/DownloadAdapter.cpp (hunk 1 above)
Comment (jamespearce2006) [inline CefSharp.Core/Internals/DownloadAdapter.cpp]: Done - I added the line to force AppVeyor to rebuild, but forgot to remove it.

**Comment by jamespearce2006** [comment]
@amaitland I agree, 2272 gives the user more control over such things. Do you think it'd be worth keeping this behavior as default, though, if the user doesn't implement the appropriate handlers?

**Comment by amaitland** [comment]
> Do you think it'd be worth keeping this behavior as default, though, if the user doesn't implement the appropriate handlers?

For downloads you have to implement `IDownloadHandler` or they're silently ignored. The default makes sense, I just think it's a little difficult to implement conditionally.

**Comment by amaitland** [comment]
I have toyed with the idea of implementing a generic set of handlers that provide some sensible defaults, make all the methods `virtual` allowing the user to override only the pieces they require. Best of both worlds in some ways, if you require the flexibility you can implement all the low lying stuff yourself, if you only need to implement one method on `IRequestHandler` then it safes a lot of effort (specially in the fact that you have to `return false` to implement the default behavior in a lot of cases). Just one of those low priority ideas that I'd never gotten around too.

**Comment by fabionardelli1982** [comment]
I checked the issue with the version 89.0.17 and the blank popup is not closed at the end of the Download of file. Then the issue is still reproducible

**Comment by amaitland** [comment]
To give users full control this change was reverted as there was no way to disable the default behaviour. 

The relevant APIs for closing the popup were added years ago so you can easily close the popup in your IDownloadHandler implementation.

**Pull request merged**